src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs

```diff
@@ -76,296 +67,106 @@ public override async Task Dispatch(PeerInfo bestPeer, BlocksRequest? blocksRequ
-            }
-            _previousBestPeer = bestPeer;
-
-            try
-            {
-                InvokeEvent(new SyncEventArgs(bestPeer.SyncPeer, Nethermind.Synchronization.SyncEvent.Started));
-                await DownloadBlocks(bestPeer, blocksRequest, cancellation)
-                        .ContinueWith(t => HandleSyncRequestResult(t, bestPeer), cancellation);
-            }
-            finally
-            {
-                _allocationWithCancellation.Dispose();
-            }
+            await base.Dispatch(bestPeer, blocksRequest, cancellation);
         }
 
         public override async Task<long> DownloadBlocks(PeerInfo? bestPeer, BlocksRequest blocksRequest,
             CancellationToken cancellation)
         {
+            // Note: Redundant with Dispatch, but test uses it.
             if (_beaconPivot.BeaconPivotExists() == false && _poSSwitcher.HasEverReachedTerminalBlock() == false)
             {
                 if (_logger.IsDebug)
                     _logger.Debug("Using pre merge block downloader");
-                return await base.DownloadBlocks(bestPeer, blocksRequest, cancellation);
+                return await _preMergeBlockDownloader.DownloadBlocks(bestPeer, blocksRequest, cancellation);
             }
 
-            if (bestPeer is null)
-            {
-                string message = $"Not expecting best peer to be null inside the {nameof(BlockDownloader)}";
-                if (_logger.IsError) _logger.Error(message);
-                throw new ArgumentNullException(message);
-            }
-
-            DownloaderOptions options = blocksRequest.Options;
-            bool downloadReceipts = (options & DownloaderOptions.WithReceipts) == DownloaderOptions.WithReceipts;
-            bool shouldProcess = (options & DownloaderOptions.Process) == DownloaderOptions.Process;
-            bool shouldMoveToMain = (options & DownloaderOptions.MoveToMain) == DownloaderOptions.MoveToMain;
-
-            int blocksSynced = 0;
-            long currentNumber = _blockTree.BestKnownNumber;
-            if (_logger.IsDebug)
-                _logger.Debug(
-                    $"MergeBlockDownloader GetCurrentNumber: currentNumber {currentNumber}, beaconPivotExists: {_beaconPivot.BeaconPivotExists()}, BestSuggestedBody: {_blockTree.BestSuggestedBody?.Number}, BestKnownNumber: {_blockTree.BestKnownNumber}, BestPeer: {bestPeer}, BestKnownBeaconNumber {_blockTree.BestKnownBeaconNumber}");
+            return await base.DownloadBlocks(bestPeer, blocksRequest, cancellation);
+        }
 
-            bool HasMoreToSync(out BlockHeader[]? headers, out int headersToRequest)
+        public override async Task<long> DownloadHeaders(PeerInfo? bestPeer, BlocksRequest blocksRequest,
+            CancellationToken cancellation)
+        {
+            // Note: Redundant with Dispatch, but test uses it.
+            if (_beaconPivot.BeaconPivotExists() == false && _poSSwitcher.HasEverReachedTerminalBlock() == false)
```

Comment:
Can we at least move this condition to a method or property? It is duplicated multiple times.

Reply:
Moved to a method